src/org/rascalmpl/uri/URIResolverRegistry.java

```diff
@@ -57,6 +59,14 @@ public class URIResolverRegistry {
 	private final Map<String, ISourceLocationWatcher> watchers = new ConcurrentHashMap<>();
 	private final Map<ISourceLocation, Set<Consumer<ISourceLocationChanged>>> watching = new ConcurrentHashMap<>();
 
+	// we allow the user to define (using -Drascal.fallbackResolver=fully.qualified.classname) a single class that will handle
+	// scheme's not statically registered. That class should implement at least one of these interfaces
+	private final AtomicReference<@Nullable ISourceLocationInput> fallbackInputResolver = new AtomicReference<>();
+	private final AtomicReference<@Nullable ISourceLocationOutput> fallbackOutputResolver = new AtomicReference<>();
+	private final AtomicReference<@Nullable ILogicalSourceLocationResolver> fallbackLogicalResolver = new AtomicReference<>();
+	private final AtomicReference<@Nullable IClassloaderLocationResolver> fallbackClassloaderResolver = new AtomicReference<>();
+	private final AtomicReference<@Nullable ISourceLocationWatcher> fallbackWatcher = new AtomicReference<>();
```

Comment:
Why are there so many fallbacks? I thought we were doing one fallback class which had to implement all the other interfaces (optionally)?

Reply:
Well, I mirrored the service loading logic, at loading time we do a single pass of instances of, and then we work with that.

I could switch the logic around to keep a single instance, and just at the use site do a instanceof check everytime?

Reply:
Never mind. I see that helps the clarity of the code further down.

Reply:
yeah, that would remove the complexity in the fields and add it to the methods. So it's just communicating vessels of a little complexity. I think your solution is more elegant.